File: src/main/java/systems/conduit/main/api/Entity.java
package systems.conduit.main.api;

import net.minecraft.core.Direction;
import net.minecraft.core.Position;
import net.minecraft.network.chat.Component;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.tags.Tag;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.Pose;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.material.Fluid;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.scores.Team;

import java.util.*;
import java.util.function.Predicate;

/**
 * Base Conduit interface for use with mixins.
 * Implementation: {@link systems.conduit.main.mixins.api.EntityMixin}
 *
 * @since API 0.1
 */
public interface Entity {

    UUID getUUID();

    double getX();
    double getY();
    double getZ();

    void sendMessage(String message);
    void sendMessage(Component message);

    Level getLevel();

    void teleport(double x, double y, double z);
    void teleport(Position position);
    void teleport(Entity entity);

    boolean isSpectator();
    void unRide();
    EntityType<?> getType();
    int getId();
    void setId(int id);
    Set<String> getTags();
    boolean addTag(String tag);
    boolean removeTag(String tag);
    void kill();
    Pose getPose();
    int getPortalWaitTime();
    void setSecondsOnFire(int time);
    void setRemainingFireTicks(int ticks);
    int getRemainingFireTicks();
    void clearFire();
    boolean isSilent();
    void setSilent(boolean silent);
    boolean isNoGravity();
    void setNoGravity(boolean gravity);
    boolean isInWater();
    boolean isInWaterOrRain();
    boolean isInWaterRainOrBubble();
    boolean isInWaterOrBubble();
    boolean isUnderWater();
    boolean isUnderLiquid(Tag<Fluid> fluid);
    boolean isInLava();
    void moveRelative(float scale, Vec3 vector);
    void moveTo(double x, double y, double z, float xRot, float yRot);
    float distanceTo(Entity entity);
    double distanceToSqr(double x, double y, double z);
    double distanceToSqr(Entity entity);
    double distanceToSqr(Vec3 vector);
    Vec3 getEyePosition(float scale);
    boolean isAlive();
    boolean isInWall();
    boolean startRiding(Entity entity);
    void stopRiding();
    boolean isSwimming();
    void setSwimming(boolean state);
    boolean isSprinting();
    void setSprinting(boolean state);
    boolean isGlowing();
    void setGlowing(boolean state);
    boolean isInvisible();
    Optional<Team> conduit_getTeam();
    boolean isAlliedTo(Team team);
    boolean isAlliedTo(Entity entity);
    void setInvisible(boolean state);
    int getMaxAirSupply();
    void setAirSupply(int supply);
    Direction getDirection();
    Optional<Entity> conduit_getControllingPassenger();
    List<Entity> conduit_getPassengers();
    boolean hasPassenger(Entity entity);
    boolean hasPassenger(Predicate<net.minecraft.world.entity.Entity> entity);  // TODO: Convert to Conduit entity
    Iterable<net.minecraft.world.entity.Entity> getIndirectPassengers();  // TODO: Convert to Conduit entity
    boolean hasExactlyOnePlayerPassenger();
    boolean isPassengerOfSameVehicle(Entity entity);
    Vec3 position();

    SoundEvent conduit_getSwimSound();
    SoundEvent conduit_getSwimSplashSound();
    SoundEvent conduit_getSwimHighSpeedSplashSound();
}

File: src/main/kotlin/systems/conduit/api/Entity.kt
package systems.conduit.api

import net.minecraft.core.Direction
import net.minecraft.core.Position
import net.minecraft.network.chat.Component
import net.minecraft.sounds.SoundEvent
import net.minecraft.tags.Tag
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.Pose
import net.minecraft.world.level.material.Fluid
import net.minecraft.world.phys.Vec3
import net.minecraft.world.scores.Team
import java.util.*


/**
 * Base Conduit interface for use with mixins.
 * Implementation: [systems.conduit.core.mixins.api.EntityMixin]
 *
 * @since API 0.1
 */
interface Entity {
    fun getUUID(): UUID?

    fun getX(): Double
    fun getY(): Double
    fun getZ(): Double

    fun sendMessage(message: String?)
    fun sendMessage(message: Component?)

    fun getLevel(): Level?

    fun teleport(x: Double, y: Double, z: Double)
    fun teleport(position: Position?)
    fun teleport(entity: Entity?)

    fun isSpectator(): Boolean
    fun unRide()
    fun getType(): EntityType<*>?
    fun getId(): Int
    fun setId(id: Int)
    fun getTags(): Set<String?>?
    fun addTag(tag: String?): Boolean
    fun removeTag(tag: String?): Boolean
    fun kill()
    fun getPose(): Pose?
    fun getPortalWaitTime(): Int
    fun setSecondsOnFire(time: Int)
    fun setRemainingFireTicks(ticks: Int)
    fun getRemainingFireTicks(): Int
    fun clearFire()
    fun isSilent(): Boolean
    fun setSilent(silent: Boolean)
    fun isNoGravity(): Boolean
    fun setNoGravity(gravity: Boolean)
    fun isInWater(): Boolean
    fun isInWaterOrRain(): Boolean
    fun isInWaterRainOrBubble(): Boolean
    fun isInWaterOrBubble(): Boolean
    fun isUnderWater(): Boolean
    fun isUnderLiquid(fluid: Tag<Fluid?>?): Boolean
    fun setInLava()
    fun isInLava(): Boolean
    fun moveRelative(scale: Float, vector: Vec3?)
    fun setLevel(level: Level?)
    fun moveTo(x: Double, y: Double, z: Double, xRot: Float, yRot: Float)
    fun distanceTo(entity: Entity?): Float
    fun distanceToSqr(x: Double, y: Double, z: Double): Double
    fun distanceToSqr(entity: Entity?): Double
    fun distanceToSqr(vector: Vec3?): Double
    fun getEyePosition(scale: Float): Vec3?
    fun isAlive(): Boolean
    fun isInWall(): Boolean
    fun startRiding(entity: Entity?): Boolean
    fun stopRiding()
    fun isSwimming(): Boolean
    fun setSwimming(state: Boolean)
    fun isSprinting(): Boolean
    fun setSprinting(state: Boolean)
    fun isGlowing(): Boolean
    fun setGlowing(state: Boolean)
    fun isInvisible(): Boolean
    fun conduit_getTeam(): Optional<Team?>?
    fun isAlliedTo(team: Team?): Boolean
    fun isAlliedTo(entity: Entity?): Boolean
    fun setInvisible(state: Boolean)
    fun getMaxAirSupply(): Int
    fun setAirSupply(supply: Int)
    fun getDirection(): Direction?
    fun conduit_getControllingPassenger(): Optional<Entity?>?
    fun conduit_getPassengers(): List<Entity?>?
    fun hasPassenger(entity: Entity?): Boolean
    fun hasPassenger(entity: Class<out net.minecraft.world.entity.Entity?>?): Boolean // TODO: Convert to Conduit entity

    fun getIndirectPassengers(): Collection<net.minecraft.world.entity.Entity?>? // TODO: Convert to Conduit entity

    fun hasOnePlayerPassenger(): Boolean
    fun isPassengerOfSameVehicle(entity: Entity?): Boolean
    fun hasIndirectPassenger(entity: Entity?): Boolean
    fun position(): Vec3?

    fun conduit_getSwimSound(): SoundEvent?
    fun conduit_getSwimSplashSound(): SoundEvent?
    fun conduit_getSwimHighSpeedSplashSound(): SoundEvent?
}
